052-connector-state.md

```diff
@@ -0,0 +1,77 @@
+
+# Support stopping connectors
+
+This proposal aims at adding support for stopping Kafka Connect connectors.
+
+## Current situation
+
+At the moment Strimzi allows pausing connectors. This is done by a property, `pause`, in the `KafkaConnectorSpec` schema. Whenever it is set to `true`, the connectors and its tasks are paused.
+
+For example:
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  tasksMax: 1
+```
+
+## Motivation
+
+Since Kafka 3.5, Kafka Connect connectors can also be stopped. This is done by using a new REST endpoint `PUT /connectors/{connector}/stop`. Compared to paused where the connector and tasks stay instantiated, when stopped only the configuration of the connector is kept but nothing is actually running.
+
+A connector paused still uses some resources but it's faster to resume. So using pause is great if you want to halt data flowing for a short duration. For longer durations it's interesting to fully stop the connector with the new API. Also the stopped state will be required to use the offset delete/reset endpoints that should come with Kafka 3.6. For these reasons, Strimzi should support both pausing and stopping connectors.
+
+The `PUT /connectors/{connector}/resume` REST endpoint is used to restart both paused and stopped connectors.
+
+## Proposal
+
+Since a connector can't be paused and stopped at the same time, the proposal is to replace the current `pause` property and add a new property, `state`, in the `AbstractConnectorSpec` schema. That new property will accept the `pause`, `stop` or `run` values.
+
+In the next Strimzi release the `pause` property will be marked as deprecated. In the following release, the `pause` field will be deleted.
+
+This proposal does not intend to address deleting/resetting offsets. As explained in the motivation section, stopping connectors has value on its own. Adding support for deleting/resetting connector offsets will be tackled in a separate proposal once Strimzi adopts a Kafka version that supports this feature (expected to release in 3.6).
+
+## Affected/not affected projects
+
+This only affects `strimzi-kafka-operator`. This will be usable by regular connectors (via `KafkaConnectorSpec`) and by MirrorMaker connectors  (via `KafkaMirrorMaker2ConnectorSpec`).
+
+## Compatibility
+
+When first upgrading to a Strimzi version that adds `state`, `state` will be set to `pause` if `pause` is `true`, otherwise `state` will default to `run`. For example, upgrading with a connector like 
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  tasksMax: 1
+```
+will result in:
+```yaml
+spec:
+  class: org.apache.kafka.connect.file.FileStreamSourceConnector
+  config:
+    file: /opt/kafka/LICENSE
+    topic: my-topic
+  pause: true
+  state: pause
+  tasksMax: 1
+```
```

Comment:
We do not really do it this way as there is no simple way to modify the resource without risks to just end up in a loop with some other tools. So normally you would do something like this:
* The new field is preferred when both are set to some value
* When the old field is set in parallel with the new field, you will issue a warning about a conflcit
* When the new field is not set (i.e. is null), it will mean `run` if `pause == false` or if `pause` is not set. Or it will equal `pause` if `pause==true`. But this will not be done as a change to the custom resource. You will just handle it that way in the code.

This might stay like this until the next APi version which removes the `pause` field. Or you can say that both fields will be supported e.g. until Strimzi 0.40. In which case, in 0.41, we can change the code to completely ignore the `pause` field in the CRs ... but we still need to keep it in the custom resource definition as deprecated.